controllers/storagecluster/provider_server.go

```diff
@@ -0,0 +1,243 @@
+		}
+	}
+
+	if finalErr == nil {
+		r.Log.Info("nigoyal Resource deletion for provider succeeded")
+	}
+
+	return finalErr
+}
+
+func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredDeployment := GetProviderApiServerDeployment(instance)
+	actualDeployment := &appsv1.Deployment{}
+
+	actualDeployment.ObjectMeta = desiredDeployment.ObjectMeta
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualDeployment,
+		func() error {
+			actualDeployment.Spec = desiredDeployment.Spec
+			return controllerutil.SetOwnerReference(instance, actualDeployment, r.Client.Scheme())
+		},
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.Log.Error(err, "nigoyal Failed to create/update deployment", "Name", desiredDeployment.Name)
+		return err
+	}
+
+	err = o.isDeploymentReady(actualDeployment, desiredDeployment)
+	if err != nil {
+		r.Log.Error(err, "nigoyal Deployment is not ready", "Name", desiredDeployment.Name)
+		return err
+	}
+
+	r.Log.Info("nigoyal Deployment is running as desired")
+	return nil
+}
+
+func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredService := GetProviderApiServerService(instance)
+	actualService := &corev1.Service{}
+
+	actualService.ObjectMeta = desiredService.ObjectMeta
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualService,
+		func() error {
+			actualService.Spec = desiredService.Spec
+			return controllerutil.SetOwnerReference(instance, actualService, r.Client.Scheme())
+		},
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.Log.Error(err, "nigoyal Failed to create/update service", "Name", desiredService.Name)
+		return err
+	}
+
+	r.Log.Info("nigoyal Service create/update succeeded")
+	return nil
+}
+
+func (o *ocsProviderServer) createSecret(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredSecret := GetProviderApiServerSecret(instance)
+	actualSecret := &corev1.Secret{}
+
+	actualSecret.ObjectMeta = desiredSecret.ObjectMeta
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualSecret,
+		func() error {
+			actualSecret.StringData = desiredSecret.StringData
+			return controllerutil.SetOwnerReference(instance, actualSecret, r.Client.Scheme())
+		},
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.Log.Error(err, "nigoyal Failed to create/update deployment", "Name", desiredSecret.Name)
+		return err
+	}
+
+	r.Log.Info("nigoyal Secret create/update succeeded")
+	return nil
+}
+
+func (o *ocsProviderServer) isDeploymentReady(actual, desired *appsv1.Deployment) error {
+
+	if actual.Status.AvailableReplicas != *desired.Spec.Replicas {
+		return fmt.Errorf("Deployment %s is not ready", desired.Name)
+	}
+
+	return nil
+}
+
+func GetProviderApiServerDeployment(instance *ocsv1.StorageCluster) *appsv1.Deployment {
+
+	return &appsv1.Deployment{
```

Comment:
I think this is a good idea.

We have the same policy in the ocs-osd-deployer for most of the things that we reconcile. We do not use `embed` but `createOrUpdate` with a mutate function that starts with the template and applies dynamic changes on top of it. But the idea is almost identical.

Reply:
I'm not convinced that using `embed` would be an actual benefit to us... but I could be convinced otherwise.

Specifically, if we move all our static definitions into YAML files, we would then be in a place where we're parsing this text blob into a go struct. My main concern is how easy is it to debug errors? For example, what happens if we've changed a field or entire sub-struct in the go definition but the YAML manifest is not longer valid? Will that be any easier than continuing to hit nil-pointer errors because we keep using pointers to structs?

Reply:
I would argue that YAML files are useless, the way we solved it in the deployer is by using const structs for the templates, one example can be found here: https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/templates/alertmanagerconfig.go

This bring the benefits of static typing, and the ability to use code to generate some of the static parts

Reply:
To complete the picture, the pattern we use for the reconcile is:
```
_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfig, func() error {
	if err := r.own(r.alertmanagerConfig); err != nil {
		return err
	}
	desired := templates.AlertmanagerConfigTemplate.DeepCopy()
	r.alertmanagerConfig.Spec = desired.Spec
	
	// Some code to apply the none static part of the reconcile
}

```

Reply:
Yeah, if we can let's do this. :point_up:

Reply:
We are doing something like this.

Reply:
That is not the same:
1. The template is kept in memory but live for all reconciles not a single one
2. A template (as an env variable or a file) is more declarative by nature, as it is a definition a declaration of a definition,  while code  (as in a func) is less unless the entire func is a declaration of a variable that is immediately returned. But at that point, there is no reason they have the function in the first place.
3. If you have a function, you do not need a template as a function can take the actual object and mutate it, but then you lose the declarative nature of templates. 


To summarize, my opinion, on the matter:
- Copy from a template is a good solution
- Mutating using code is also a good solution (but less readable and prune to ordering errors) 
- Combining them is does NOT create something that is better then the original options

Reply:
I'm happy with this: https://github.com/red-hat-storage/ocs-operator/pull/1421/commits/bb44cbf3b72380f6aa1013d1c1d2834155e05d1e#diff-85e662b500fee0a341fbae01f02cd05d31663d6f6e355ab6536eb8b2ca0ea38fR91-R101